crates/rattler_cache/src/package_cache/mod.rs

```diff
@@ -98,17 +131,151 @@ impl From<Cancelled> for PackageCacheError {
+            .unwrap_or(false)
+    }
+
+    /// Validate the package, and throw if invalid.
+    pub async fn validate_or_throw(
+        &self,
+        cache_key: &CacheKey,
+    ) -> Result<CacheLock, PackageCacheLayerError> {
+        let cache_entry = self
+            .packages
+            .get(&cache_key.clone().into())
+            .ok_or(PackageCacheLayerError::PackageNotFound)?
+            .clone();
+        let mut cache_entry = cache_entry.lock().await;
+        let cache_path = self.path.join(cache_key.to_string());
+
+        match validate_package_common::<
+            fn(PathBuf) -> _,
+            Pin<Box<dyn Future<Output = Result<(), _>> + Send>>,
+            std::io::Error,
+        >(
+            cache_path,
+            cache_entry.last_revision,
+            cache_key.sha256.as_ref(),
+            None,
+            None,
+        )
+        .await
+        {
+            Ok(cache_lock) => {
+                cache_entry.last_revision = Some(cache_lock.revision);
+                cache_entry.last_sha256 = cache_lock.sha256;
+                Ok(cache_lock)
+            }
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Validate the package, and fetch it if invalid.
+    pub async fn validate_or_fetch<F, Fut, E>(
+        &self,
+        fetch: F,
+        cache_key: &CacheKey,
+        reporter: Option<Arc<dyn CacheReporter>>,
+    ) -> Result<CacheLock, PackageCacheLayerError>
+    where
+        F: (Fn(PathBuf) -> Fut) + Send + 'static,
+        Fut: Future<Output = Result<(), E>> + Send + 'static,
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        let entry = self
+            .packages
+            .entry(cache_key.clone().into())
+            .or_default()
+            .clone();
+
+        let mut cache_entry = entry.lock().await;
+        let cache_path = self.path.join(cache_key.to_string());
+
+        match validate_package_common(
+            cache_path,
+            cache_entry.last_revision,
+            cache_key.sha256.as_ref(),
+            Some(fetch),
+            reporter,
+        )
+        .await
+        {
+            Ok(cache_lock) => {
+                cache_entry.last_revision = Some(cache_lock.revision);
+                cache_entry.last_sha256 = cache_lock.sha256;
+                Ok(cache_lock)
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
 impl PackageCache {
-    /// Constructs a new [`PackageCache`] located at the specified path.
+    /// Constructs a new [`PackageCache`] with only one layer.
     pub fn new(path: impl Into<PathBuf>) -> Self {
-        Self {
-            inner: Arc::new(PackageCacheInner {
+        Self::new_layered(std::iter::once(path.into()))
+    }
+
+    /// Constructs a new [`PackageCache`] located at the specified paths.
+    /// Read-only layers are queried first.
+    /// Within read-only layers, the ordering is defined in this constructor. Ditto for writable layers.
```

Comment:
I dont fully understand this comment and I also dont see how it relates to the code? To me it makes most sense to pass in the paths in the order the user wants the layers to be ordered.

I would assume you look through the layers in the order they are passed in, and only write to the first writable layer.

Reply:
Seems like I didn't make the comment clear. What I meant was, the querying order is defined by 
- Primary key: read-only before write-only
- Secondary key: order it was passed into the constructor

But now that I think about it, querying all layers in the order passed into the constructor makes more sense.